extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,491 @@
+			tenantId: string
+			displayName: string
+		}
+		const allSubs: Subscription[] = [];
+		const tokens = await this.getSecurityToken(account, azdata.AzureResource.ResourceManagement);
+
+		for (const tenant of account.properties.tenants) {
+			const token = tokens[tenant.id];
+			const subscriptionUri = url.resolve(this.metadata.settings.armResource.endpoint, 'subscriptions?api-version=2019-11-01');
+			try {
+				const subscriptionResponse = await this.makeGetRequest(token.token, subscriptionUri);
+				const subscriptions: Subscription[] = subscriptionResponse.data.value.map((subscriptionInfo: SubscriptionResponse) => {
+					return {
+						id: subscriptionInfo.subscriptionId,
+						displayName: subscriptionInfo.displayName,
+						tenantId: subscriptionInfo.tenantId
+					} as Subscription;
+				});
+				allSubs.push(...subscriptions);
+			} catch (ex) {
+				console.log(ex);
+				throw new Error('Error retreiving subscription information');
+			}
+		}
+		return allSubs;
+	}
+
+	protected async getToken(postData: { [key: string]: string }, tenant = this.commonTenant, resourceId: string = ''): Promise<TokenRefreshResponse | undefined> {
+		try {
+			const tokenUrl = `${this.loginEndpointUrl}${tenant}/oauth2/token`;
+
+			const tokenResponse = await this.makePostRequest(tokenUrl, postData);
+			const tokenClaims = this.getTokenClaims(tokenResponse.data.access_token);
+
+			const accessToken: AccessToken = {
+				token: tokenResponse.data.access_token,
+				key: tokenClaims.email || tokenClaims.unique_name || tokenClaims.name,
+			};
+
+			this.memdb.set(`${accessToken.key}_${tenant}_${resourceId}`, tokenResponse.data.expires_on);
+
+			const refreshToken: RefreshToken = {
+				token: tokenResponse.data.refresh_token,
+				key: accessToken.key
+			};
+
+			return { accessToken, refreshToken, tokenClaims };
+
+		} catch (err) {
+			console.dir(err);
+			const msg = localize('azure.noToken', "Retrieving the token failed.");
+			vscode.window.showErrorMessage(msg);
+			throw new Error(err);
+		}
+	}
+
+	protected getTokenClaims(accessToken: string): TokenClaims | undefined {
+		try {
+			const split = accessToken.split('.');
+			return JSON.parse(Buffer.from(split[1], 'base64').toString('binary'));
+		} catch (ex) {
+			throw new Error('Unable to read token claims');
+		}
+	}
+
+	private async refreshAccessToken(account: azdata.AccountKey, rt: RefreshToken, tenant?: Tenant, resource?: Resource): Promise<void> {
+		const postData: { [key: string]: string } = {
+			grant_type: 'refresh_token',
+			refresh_token: rt.token,
+			client_id: this.clientId,
+			tenant: this.commonTenant,
+		};
+
+		if (resource) {
+			postData.resource = resource.endpoint;
+		}
+
+		const { accessToken, refreshToken } = await this.getToken(postData, tenant?.id, resource?.id);
+
+		if (!accessToken || !refreshToken) {
+			console.log('Access or refresh token were undefined');
+			const msg = localize('azure.refreshTokenError', "Error when refreshing your account.");
```

Comment:
User error

Reply:
Hmm not this one. It's called from two places and one of them I don't want to show it to the user.

Reply:
![image](https://user-images.githubusercontent.com/1465785/77210272-ceda5600-6abd-11ea-94e9-fabfe252dd9f.png)